diff --git a/node_modules/@expo/metro-runtime/build/LoadingView.native.js b/node_modules/@expo/metro-runtime/build/LoadingView.native.js
index c7ba770..b29469c 100644
--- a/node_modules/@expo/metro-runtime/build/LoadingView.native.js
+++ b/node_modules/@expo/metro-runtime/build/LoadingView.native.js
@@ -6,7 +6,7 @@ try {
 }
 catch {
     // In react-native 0.75.0 LoadingView was renamed to DevLoadingView
-    LoadingView = require('react-native/Libraries/Utilities/DevLoadingView');
+    // LoadingView = require('react-native/Libraries/Utilities/DevLoadingView');
 }
 exports.default = LoadingView;
 //# sourceMappingURL=LoadingView.native.js.map
