// babel-cli/lib/_babel-node.js
"use strict";

var _interopRequireDefault = require("babel-runtime/helpers/interop-require-default")["default"];

var _interopRequireWildcard = require("babel-runtime/helpers/interop-require-wildcard")["default"];

var _pathIsAbsolute = require("path-is-absolute");

var _pathIsAbsolute2 = _interopRequireDefault(_pathIsAbsolute);

var _commander = require("commander");

var _commander2 = _interopRequireDefault(_commander);

var _module3 = require("module");

var _module4 = _interopRequireDefault(_module3);

var _util = require("util");

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _repl = require("repl");

var _repl2 = _interopRequireDefault(_repl);

var _babelCore = require("babel-core");

var babel = _interopRequireWildcard(_babelCore);

var _vm = require("vm");

var _vm2 = _interopRequireDefault(_vm);

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

require("babel-polyfill");

var _babelRegister = require("babel-register");

var _babelRegister2 = _interopRequireDefault(_babelRegister);

var program = new _commander2["default"].Command("babel-node");

program.option("-e, --eval [script]", "Evaluate script");
program.option("-p, --print [code]", "Evaluate script and print result");
program.option("-o, --only [globs]", "");
program.option("-i, --ignore [globs]", "");
program.option("-x, --extensions [extensions]", "List of extensions to hook into [.es6,.js,.es,.jsx]");
program.option("-w, --plugins [string]", "", _babelCore.util.list);
program.option("-b, --presets [string]", "", _babelCore.util.list);

var pkg = require("../package.json");
program.version(pkg.version);
program.usage("[options] [ -e script | script.js ] [arguments]");
program.parse(process.argv);

//

_babelRegister2["default"]({
  extensions: program.extensions,
  ignore: program.ignore,
  only: program.only,
  plugins: program.plugins,
  presets: program.presets
});

//

var replPlugin = function replPlugin() {
  return {
    visitor: {
      ModuleDeclaration: function ModuleDeclaration(path) {
        throw path.buildCodeFrameError("Modules aren't supported in the REPL");
      },

      VariableDeclaration: function VariableDeclaration(path) {
        if (path.node.kind !== "var") {
          throw path.buildCodeFrameError("Only `var` variables are supported in the REPL");
        }
      }
    }
  };
};

//

var _eval = function _eval(code, filename) {
  code = code.trim();
  if (!code) return undefined;

  code = babel.transform(code, {
    filename: filename,
    presets: program.presets,
    plugins: (program.plugins || []).concat([replPlugin])
  }).code;

  return _vm2["default"].runInThisContext(code, {
    filename: filename
  });
};

if (program.eval || program.print) {
  var code = program.eval;
  if (!code || code === true) code = program.print;

  global.__filename = "[eval]";
  global.__dirname = process.cwd();

  var _module2 = new _module4["default"](global.__filename);
  _module2.filename = global.__filename;
  _module2.paths = _module4["default"]._nodeModulePaths(global.__dirname);

  global.exports = _module2.exports;
  global.module = _module2;
  global.require = _module2.require.bind(_module2);

  var result = _eval(code, global.__filename);
  if (program.print) {
    var output = _lodash2["default"].isString(result) ? result : _util.inspect(result);
    process.stdout.write(output + "\n");
  }
} else {
  if (program.args.length) {
    (function () {
      // slice all arguments up to the first filename since they're babel args that we handle
      var args = process.argv.slice(2);

      var i = 0;
      var ignoreNext = false;
      _lodash2["default"].each(args, function (arg, i2) {
        if (ignoreNext) {
          ignoreNext = false;
          return;
        }

        if (arg[0] === "-") {
          var parsedArg = program[arg.slice(2)];
          if (parsedArg && parsedArg !== true) {
            ignoreNext = true;
          }
        } else {
          i = i2;
          return false;
        }
      });
      args = args.slice(i);

      // make the filename absolute
      var filename = args[0];
      if (!_pathIsAbsolute2["default"](filename)) args[0] = _path2["default"].join(process.cwd(), filename);

      // add back on node and concat the sliced args
      process.argv = ["node"].concat(args);
      process.execArgv.unshift(__filename);

      _module4["default"].runMain();
    })();
  } else {
    replStart();
  }
}

function replStart() {
  _repl2["default"].start({
    prompt: "> ",
    input: process.stdin,
    output: process.stdout,
    eval: replEval,
    useGlobal: true
  });
}

function replEval(code, context, filename, callback) {
  var err = undefined;
  var result = undefined;

  try {
    if (code[0] === "(" && code[code.length - 1] === ")") {
      code = code.slice(1, -1); // remove "(" and ")"
    }

    result = _eval(code, filename);
  } catch (e) {
    err = e;
  }

  callback(err, result);
}
